### ISSUE TITLE
20 second lag when adding search results to current playlist using HEAD

### ISSUE THREAD

**Issue by peterlvilim** [mopidy/mopidy-spotify#30]
I was interested in adding playlist configuration.  So I pulled head of mopidy, pyspotify, and mopidy-spotify to get started on developing.

I've run into an odd performance regression.  If I do a search in ncmpcpp (my client) the search goes fine but then when I click enter on the result (e.g. an artist name) to add those results to the current playlist ncmpcpp hangs for 20 seconds before the results are added and music starts playing.

So I checked out these commits:

mopidy-spotify:

```
commit 7197194ed455b2425c2e8ff993bf3819c15ef8ff
Author: Stein Magnus Jodal <stein.magnus@jodal.no>
Date:   Mon Jul 21 02:03:22 2014 +0200

    Release v1.2.0
```

pyspotify:

```
commit 68cdfb5fbd405c71c9ffcad5ed2647fde4d9d92e
Author: Stein Magnus Jodal <stein.magnus@jodal.no>
Date:   Sun May 4 23:22:09 2014 +0200

    docs: Release v2.0.0b3
```

And the problem goes away (e.g. items add near instantly as when using AUR version).

Had a few questions:
- Are you accepting pull requests for playlist configuration?  If so should I be developing against head?  Not sure what its current state is given that I ran into this bug and wasn't able to find anything in the bug tracker.
- Is it worth me tracking down why this is happening/have you seen this before?

Background info on my setup:
- No local music library or other plugins used besides mopidy-spotify.
- I removed everything except for my config file in ~/.config/mopidy to make sure I didn't have cache problems.
- I'm on Arch.  Packages are fully upgraded.  Was previously running mopidy-spotify from AUR.
- My ncmpcpp version is 0.6.2.  Have not tried any other clients yet.
- I did setup via sudo python2 setup.py develop for all three (mopidy, pyspotify, mopidy-spotify).  Pulled pyspotify because it looks like mopidy-spotify depends on a version not available to pip.

**Comment by jodal** [comment]
Mopidy-Spotify 1.x requires pyspotify 1.x. Mopidy-Spotify 2.x requires pyspotify 2.x.

Any new development should happen on Mopidy-Spotify 2.x.

**Comment by jodal** [comment]
To add a bit more detail: Mopidy-Spotify 2.x is close to a first release to PyPI. The current state of the develop branch is quite good. It depends on a feature of pyspotify that is not released yet, thus you need pyspotify 2 from git too.

As to adding features, discussing the feature and what it should do first greatly increases the odds for getting it merged. What kind of playlist configurations are you thinking about and what's their use cases?

It might be relevant that I've already added a new configuration in Mopidy-Spotify 2 for turning off playlists:

> spotify/allow_playlists: Whether or not playlists should be exposed. Defaults to true.

**Comment by peterlvilim** [comment]
Thanks for quick response and explanation!  Switching my branch to develop from master for mopidy-spotify solved my problem.  I have pyspotify checked out and setup also right now.

I'll comment on the open issue for playlists with what I was thinking of doing.

**Comment by peterlvilim** [comment]
I'm reopening.  The problem seems to be dependent on which artist you look up and the state of your cache, so I think I was a bit spurious in closing.

I'm using the develop branches for both mopidy-spotify and pyspotify.  Believe the problem is in mopidy-spotify.  I've opened a pull request with the area of code that I think is a problem.

A good way to illustrate the problem is to lookup an artist with a large number of albums.  I'm using ncmpcpp.  When I do a search for Avicii then select the artist to load all of the tracks for the artist into the now playing list MPD hangs for a very long time and ncmpcpp becomes unresponsive (I say very long because I haven't seen it return for over 1 minute+).  Another artist that you could do would be Selena Gomez which hangs for about 10-15 seconds but does return (fewer albums I assume).

This problem doesn't happen when I use the pre-2.x version.  I believe the problem is due to the new way that artist look ups are done is generating many queries.  Believe my diff in the pull request and the attached comments should illustrate this.

Let me know if you can reproduce.

**Comment by m0dprobe** [comment]
I'm also seeing this behaviour and I think @peterlvilim is correct in his assumption. The changes made to fix #7 seem to be a very likely suspect.

**Comment by jodal** [comment]
I can reproduce and agree with your conclusions. Closing this issue and continuing discussion in #31.
